src/fragment-generation-utils.js

```diff
@@ -298,6 +301,43 @@ const forwardTraverse = (walker, overrideMap) => {
   return walker.nextNode();
 };
 
+/**
+ * Performs backwards traversal on a TreeWalker, such that parent nodes are
+ * encountered *before* their children (except when they are ancestors of the
+ * starting node |origin|). This is useful for finding block boundaries.
+ * @param {TreeWalker} walker - the TreeWalker to be traversed
+ * @param {Set<Node>} visited - a set used to avoid repeat iterations. Should be
+ *     empty the first time this method is called.
+ * @param {Node} origin - the node where traversal started
+ * @return {Node} - |walker|'s new current node, or null if the current node
+ *     was unchanged (and thus, no further traversal is possible)
+ */
+const backwardTraverse =
+    (walker, visited, origin) => {
+      // Infinite loop to avoid recursion. Will terminate since visited set
+      // guarantees children of a node are only traversed once, and parent node
+      // will be null once the root of the walker is reached.
+      while (true) {
```

Comment:
Don't know if I missed an obvious line but, are we updating the walker's current node in this loop? If not, how do the conditions change to get the loop to stop?

Reply:
Nevermind, just looked at the TreeWalker docs.